lib/cn/cn_tree.c

```diff
@@ -1855,49 +1919,170 @@ cn_comp_cleanup(struct cn_compaction_work *w)
+    if (err)
         goto done;
 
-    /* Punt if job on head of list is not done or another thread is already committing it. */
-    if (!atomic_read(&w->cw_rspill_done) || atomic_read(&w->cw_rspill_commit_in_progress)) {
-        w = 0;
+    err = cn_mblocks_commit(w->cw_mp, 1, mblks, CN_MUT_OTHER);
+    if (err)
+        goto done;
+
+    err = kvset_open(w->cw_tree, ss->ss_kvsetid, &km, &kvset);
+    if (err)
         goto done;
-    }
 
-    /* Job on head of spill completion list is ready to be processed.
-     * - Set "commit_in_progress" status, but leave on list until commit is done.
-     * - If the node is wedged, it means an earlier job has failed, in
-     *   which case we force failure on this job to prevent out of
-     *   order completion.
-     * - If the node is not wedged, and this job has failed then it
-     *   will cause the node to be wedged, but this will be handled
-     *   later to catch downstream errors.
+    /* CNDB: Ack kvset add record.
      */
+    err = cndb_record_kvset_add_ack(cndb, tx, cookie);
+    if (err)
+        goto done;
 
-    atomic_set(&w->cw_rspill_commit_in_progress, 1);
+    cn_comp_kvset_append(w, ss->ss_node, kvset);
 
-    if (ev(tree->ct_rspills_wedged && !w->cw_err)) {
-        w->cw_err = merr(ESHUTDOWN);
-        w->cw_canceled = true;
-    }
+    blk_list_free(&ss->ss_mblks.kblks);
+    blk_list_free(&ss->ss_mblks.vblks);
 
 done:
-    mutex_unlock(&tree->ct_rspills_lock);
+    w->cw_t4_commit = get_time_ns();
+
+    if (err) {
+        cndb_record_nak(cndb, tx);
+        kvset_put_ref(kvset);
+    }
+
+    return err;
+}
+
+static merr_t
+cn_node_apply_subspill(struct subspill *ss)
+{
+    struct cn_compaction_work *w = ss->ss_work;
+    merr_t err;
+
+    err = cn_subspill_commit(ss);
+    if (err)
+        return err;
+
+    if (ss->ss_added) {
+        w->cw_output_nodev[0] = ss->ss_node;
+        w->cw_checkpoint(w);
+    }
 
-    return w;
+    return 0;
+}
+
+static merr_t
+cn_comp_spill(struct cn_compaction_work *w)
+{
+    struct subspill *ss;
+    struct spillctx *sctx;
+    struct cn_tree *tree = w->cw_tree;
+    struct route_node *rtn = 0;
+    uint cnum = 0;
+    merr_t err;
+
+    err = cn_spill_init(w, &sctx);
+    if (err)
+        return err;
+
+    while (!err) {
+        struct cn_tree_node *node;
+        void *lock;
+        struct kvset_list_entry *le;
+        uint64_t node_dgen;
+        unsigned char ekey[HSE_KVS_KEY_LEN_MAX];
+        uint eklen;
+
+        rmlock_rlock(&tree->ct_lock, &lock);
+
+        rtn = rtn ? route_node_next(rtn) : route_map_first_node(tree->ct_route_map);
+        if (!rtn) {
+            rmlock_runlock(lock);
+            break;
+        }
+
+        node = route_node_tnode(rtn);
+
+        le = list_first_entry_or_null(&node->tn_kvset_list, typeof(*le), le_link);
+        node_dgen = le ? kvset_get_dgen(le->le_kvset) : 0;
+
+        route_node_keycpy(rtn, ekey, sizeof(ekey), &eklen);
+        rmlock_runlock(lock);
```

Comment:
Try not to change any of this too much if you can help it, my new spill/split sync stuff merges into this code above :-)